### PR TITLE
[API] Deliver ABI breaking changes

### LINKED ISSUE #1983: Prepare release, include fixes with abi breaking changes
Prepare future releases with breaking changes.

Due to expected changes in the spec, there will be breaking changes in the API/ABI.

Due to changes in the implementation, there are also known changes in the SDK interface, which can break applications when using shared libraries.

It is desirable to group many fixes that causes breaking changes at once, so listing here all the known issues that impact API/ABI.

It is also desirable to define the overall strategy to introduce breaking changes.

API/ABI:

* #2033 (API + SDK)
* #2032 (API + SDK)
* #1922 (API)
* #1416 (API)
* #828 (API) -> Providing alternate Span creation API to return nostd::unique_ptr.
* #2155
* #1753
* #2132
* #2279

SDK:

* https://github.com/open-telemetry/opentelemetry-cpp/issues/1845 (remove exporter options struct inheritance)
* https://github.com/open-telemetry/opentelemetry-cpp/issues/1402 (new members in exporter options struct)

### PR DESCRIPTION
**Pull request by marcalff**
Fixes #1983

## Changes

Adjust the makefiles to support multiple ABIs.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by owent — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -92,17 +92,30 @@ if(VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
...
+option(WITH_ABI_VERSION_1 "ABI version 1" ON)
+option(WITH_ABI_VERSION_2 "EXPERIMENTAL: ABI version 2 preview" OFF)
+
 file(READ "${CMAKE_CURRENT_LIST_DIR}/api/include/opentelemetry/version.h"
      OPENTELEMETRY_CPP_HEADER_VERSION_H)
-if(OPENTELEMETRY_CPP_HEADER_VERSION_H MATCHES
-   "OPENTELEMETRY_ABI_VERSION_NO[ \t\r\n]+\"?([0-9]+)\"?")
-  math(EXPR OPENTELEMETRY_ABI_VERSION_NO ${CMAKE_MATCH_1})
+
+if(WITH_ABI_VERSION_2)
```
Comment (ThomsonTan) [inline CMakeLists.txt:101]: Error out if more than one ABI versions are set?

Comment (marcalff) [reply CMakeLists.txt:126]: Good point, thanks.

Fixed.

**Comment by marcalff** [comment]
Blocking, prerequisite for:
* #2224
* ABI breaking changes in general

**Comment by lalitb** [comment]
Still looking into the PR. Will be completing it by tomorrow :)

File: api/CMakeLists.txt [hunk 2]
```diff
@@ -94,6 +94,14 @@ if(WITH_REMOVE_METER_PREVIEW)
                              INTERFACE ENABLE_REMOVE_METER_PREVIEW)
 endif()
 
+if(WITH_ABI_VERSION_2)
+  target_compile_definitions(opentelemetry_api
+                             INTERFACE OPENTELEMETRY_ABI_VERSION_NO=2)
+elseif(WITH_ABI_VERSION_1)
+  target_compile_definitions(opentelemetry_api
+                             INTERFACE OPENTELEMETRY_ABI_VERSION_NO=1)
+endif()
```
[the remark this answers is not in the document]
Comment (marcalff) [reply api/CMakeLists.txt]: Fixed.

File: docs/abi-version-policy.md [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+
+Instrumented applications are built against ABI v1 by default,
+but may ask to use ABI v2 instead.
+
+opentelemetry-cpp produces a library for ABI v1 by default,
+but can be configured to provide ABI v2 instead.
+
+Fixes introducing breaking changes can not be delivered ABI v2, because it
+is declared stable. These fixes can either wait, or an experimental ABI v3
+can be created.
```
[the remark this answers is not in the document]
Comment (marcalff) [reply docs/abi-version-policy.md]: Yes. Reworded for clarity.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for the work.

**Pull request merged**